### DISCUSSION TITLE
Path naming

### DISCUSSION THREAD

**Discussion by xstrex** [beetbox/beets#5144] [Q&A]
Currently I've got a music directory which contains a mix of 'artist' folders, and 'albumartist' folders, and frankly it's a mess. 

As an example here's a few folder names:
deadmau5
deadmau5 & Chris Lake
deadmau5 & Kaskade
deadmau5 & Kiesza
deadmau5 & Lights

I'm wondering if there's a better way to create folders for albums with multiple artists; instead of having a separate folder for each of them on the root of the music directory? 

I've read the [note](https://beets.readthedocs.io/en/stable/reference/pathformat.html#a-note-about-artists) about artists, and the logic makes sense. But it also doesn't make sense to me to have a folder at the root of the music directory each time 2 artist get together and make an album. 

My current default path is: $albumartist/$album%aunique{}/$albumartist - $album - $track - $title
( I name the tracks as such in case they get copied to a single directory, they will still sort properly by artist, album, track.. )

In the [example](https://musicbrainz.org/release/798dcaab-0f1a-4f02-a9cb-61d5b0ddfd36.html) in the documentation it makes sense how a single album could be split into multiple root level artist folders; I guess that's why this feels so messy to me. 

Is there a template I'm unaware of for 'release artist' that would always be deadmau5 since that's the artist who released the album, even though the album contains tracks from multiple artists? Feel like I'm not fully grasping this, but I hope that makes sense.

**Comment by DjSlash** [comment]
I think these collaborations are mostly producing singles. Since you can use queries for your paths configuration **and** use combinations, you could do something in the likes of:

```yaml
paths:
    albumartist::deadmau5 albumtype:single deadmau5/$album%aunique{}/$albumartist - $album - $track - $title
```

It should match albumartist that contains "deadmau5" (as a regexp) that are singles. 

Plugins that might be of help:
- [rewrite](https://beets.readthedocs.io/en/stable/plugins/rewrite.html)
- [subsitute](https://beets.readthedocs.io/en/stable/plugins/substitute.html)

**Reply by xstrex** [reply]
Thank you for your help on this; though I'm not exactly sure how to implement it. deadmau5 is just one example, and in reality I have 325 folders in my music directory containing multi-artist singles, so my understanding is that I would have to make a unique `paths:` statement for each of them, and what happens when a new artist gets added? Even if I used `substitute` would I have to make a substitute line item for each and every artist? Am I misunderstanding something?

**Reply by DjSlash** [reply]
You understood correctly. However, in the recently released [2.2.0](https://github.com/beetbox/beets/releases/tag/v2.2.0), there is a change for the subsitute plugin. It allows to group things together with a more generic rule.

**Comment by xstrex** [comment]
Alright, so I'm trying to use the new improved `substitute` plugin, but running into some interesting issues. 
First the config:
My plugins list does contain `substitute`

My path statement:
`default: %substitute{$albumartist}/$album%aunique{}/$albumartist - $album - $track - $title`
_I've tried starting the function call with a `\` as in the docs, but that didn't work, since a `replace:` statement was searching for items starting with a `\`_

My actual substitute config:
```yml
substitute:
  ^(.*?)(,| &| and| +| feat.| vs).*: \1
```

I just ran an import on an existing item, appending the `-v` to beet, and `-A` to import, verbose ans skip tagging (since the item is already tagged). On this directory in the library: `/music/Ace Ventura & Freedom Fighters`
Which should catch since it contains an `&`, the command completed successfully, but now I can't find the album.. Here's the verbose output:
```
user configuration: /config/config.yaml
data directory: /config
plugin paths:
fetchart: google: Disabling art source due to missing key
fetchart: lastfm: Disabling art source due to missing key
lastgenre: Loading canonicalization tree /lsiopy/lib/python3.12/site-packages/beetsplug/lastgenre/genres-tree.yaml
lyrics: Disabling google source: no API key configured.
Sending event: pluginload
library database: /config/library.db
library directory: /music
Sending event: library_opened
Sending event: import_begin
Sending event: import_task_created
/music/Ace Ventura & Freedom Fighters/The Encounter
Replacing item 18830: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 01 - The Encounter (original mix).m4a
Sending event: database_change
Sending event: item_removed
Replacing item 18831: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 02 - The Encounter (LOUD remix).m4a
Sending event: database_change
Sending event: item_removed
Replacing item 18832: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 03 - The Encounter (Ryanosaurus remix).m4a
Sending event: database_change
Sending event: database_change
Sending event: album_removed
Sending event: item_removed
3 of 3 items replaced
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Reimported album 6388. Preserving attribute ['added']. Path: /music/Ace Ventura & Freedom Fighters/The Encounter
Reimported album 6388. Preserving flexible attributes []. Path: /music/Ace Ventura & Freedom Fighters/The Encounter
Reimported item 50190. Preserving attribute ['added']. Path: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 01 - The Encounter (original mix).m4a
Reimported item 50190. Preserving flexible attributes ['track_alt', 'data_source']. Path: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 01 - The Encounter (original mix).m4a
Sending event: database_change
Reimported item 50191. Preserving attribute ['added']. Path: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 02 - The Encounter (LOUD remix).m4a
Reimported item 50191. Preserving flexible attributes ['track_alt', 'data_source']. Path: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 02 - The Encounter (LOUD remix).m4a
Sending event: database_change
Reimported item 50192. Preserving attribute ['added']. Path: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 03 - The Encounter (Ryanosaurus remix).m4a
Reimported item 50192. Preserving flexible attributes ['track_alt', 'data_source']. Path: /music/Ace Ventura & Freedom Fighters/The Encounter/Ace Ventura & Freedom Fighters - The Encounter - 03 - The Encounter (Ryanosaurus remix).m4a
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
lyrics: lyrics already present: Ace Ventura & Freedom Fighters - The Encounter - The Encounter (original mix)
lyrics: lyrics already present: Ace Ventura & Freedom Fighters - The Encounter - The Encounter (LOUD remix)
lyrics: lyrics already present: Ace Ventura & Freedom Fighters - The Encounter - The Encounter (Ryanosaurus remix)
Parsed query: AndQuery([NoneQuery('album_id', True)])
Parsed sort: NullSort()
Parsed query: AndQuery([BooleanQuery('comp', 1, fast=True)])
Parsed sort: NullSort()
Parsed query: AndQuery([SubstringQuery('albumtype', 'soundtrack', fast=True)])
Parsed sort: NullSort()
Sending event: before_item_moved
Sending event: item_moved
Sending event: database_change
Sending event: database_change
Parsed query: AndQuery([NoneQuery('album_id', True)])
Parsed sort: NullSort()
Parsed query: AndQuery([BooleanQuery('comp', 1, fast=True)])
Parsed sort: NullSort()
Parsed query: AndQuery([SubstringQuery('albumtype', 'soundtrack', fast=True)])
Parsed sort: NullSort()
Sending event: before_item_moved
Sending event: item_moved
Sending event: database_change
Sending event: database_change
Parsed query: AndQuery([NoneQuery('album_id', True)])
Parsed sort: NullSort()
Parsed query: AndQuery([BooleanQuery('comp', 1, fast=True)])
Parsed sort: NullSort()
Parsed query: AndQuery([SubstringQuery('albumtype', 'soundtrack', fast=True)])
Parsed sort: NullSort()
Sending event: before_item_moved
Sending event: item_moved
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: import_task_files
Sending event: write
Sending event: after_write
Sending event: write
Sending event: after_write
Sending event: write
Sending event: after_write
Sending event: album_imported
Sending event: import
Sending event: cli_exit
plexupdate: Updating Plex library...
plexupdate: ... started.
```

I've search for the filename, excluding the path, and it's now missing from the library..  

Edit; Correction, it put the files in a new directory called `Ace`, not sure why but that's where they are now. Quite confused.

**Reply by xstrex** [reply]
OK I mostly fixed my own issue, I wasn't properly escaping my added regex characters, this is working much better.. 
```yml
substitute:
  ^(.*?)(,| &| and| \+| feat\.| vs).*: \1
```

**Comment by xstrex** [comment]
One thing that I'm still running into though, in some instances it removes the empty source directory, like I'd expect. In other instances it does not remove the source directory, and leaves the file `album.nfo` behind, yet moves all the other files to the appropriate destination directory. Not really sure why that is yet.
